src/resolvers.ts

```diff
@@ -0,0 +1,130 @@
+import { Book} from "./models/book";
+import { Customer } from "./models/customer";
+import { LendingRecord } from "./models/lendingRecord";
+import { BooksDataSource } from "./dataSources/booksDataSource";
+import { CustomersDataSource } from "./dataSources/customersDataSource";
+import { logger } from "./logger";
+
+const booksDataSource = new BooksDataSource();
+const customersDataSource = new CustomersDataSource();
+
+export const resolvers = {
+  Query: {
+    // Get all books
+    books: (): Book[] => booksDataSource.getBooks(),
+
+    // Get a book by id
+    book: (_: any, { id }: { id: string }): Book | undefined => {
+      return booksDataSource.getBookById(id);
+    },
+
+    // Get all customers
+    customers: (): Customer[] => customersDataSource.getCustomers(),
+
+    // Get a customer by id
+    customer: (_: any, { id }: { id: string }): Customer | undefined =>
+      customersDataSource.getCustomerById(id),
+  },
+
+  Mutation: {
+    // Add a new book, the isLent field is set to false by default
+    addBook: (
+      _: any,
+      {title, author }: {title: string; author: string }
+    ): Book => {
+      const newBook = new Book(title, author);
+      booksDataSource.addBook(newBook);
+      return newBook;
+    },
+
+    // Add a new customer
+    addCustomer: (
+      _: any,
+      {name, email }: {name: string; email: string }
+    ): Customer => {
+      const newCustomer = new Customer(name, email);
+      customersDataSource.addCustomer(newCustomer);
+      return newCustomer;
+    },
+
+    // Lend a book to a customer, updates the fields of a book related to lending
+    lendBook: (
+      _: any,
+      {
+        bookId,
+        customerId,
+        lentDate,
+        dueDate,
+      }: {
+        bookId: string;
+        customerId: string;
+        lentDate: string;
+        dueDate: string;
+      }
+    ): Book => {
+      const book = booksDataSource.getBookById(bookId);
+      const customer = customersDataSource.getCustomerById(customerId);
+
+      if (book && customer && !book.isLent) {
+        const record = new LendingRecord(customer.id, lentDate, dueDate);
+
+        book.isLent = true;
+        book.currentLendeeId = customer.id;
+        book.lendingHistory.push(record);
+        booksDataSource.updateBook(book);
+
+        return book;
+      } else {
+        const errorMessage = `Book with ID ${bookId} is already lent or does not exist.`;
```

Comment:
This is too much of an ambiguous message. I would suggest handling these separately.

Reply:
After following the "keep the happy path on the left side" idea all the error messages are more specific.